Issue/31562 filter coupons by status

anitaa1990 opened the pull request:
Closes #31562

## All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

## Background

The Woo mobile app is working on adding coupon management to the app. As part of that project, there are a few enhancements in the GET Coupon(s) API endpoint that we would need to make. 
- It looks like only published coupons are available in the API response. It would be nice to list all coupons available for the store, regardless of the status, since Core displays it that way. 

### Designs
<img width="1541" alt="Screenshot 2022-01-05 at 4 03 09 PM" src="https://user-images.githubusercontent.com/22608780/148203152-e532f716-8538-4ea2-a85a-e8267d6c329b.png">
<img width="292" alt="Screenshot 2022-01-05 at 4 03 52 PM" src="https://user-images.githubusercontent.com/22608780/148203270-9a1112a5-842f-4e3d-9cfb-152bc282da84.png">

## Proposed Changes
This PR adds the option to filter coupons by status (if requested in the API request). Using an example JSON, this is how it would look like:
- GET `/wc/v3/coupons` would return all coupons, regardless of the status. The default statuses I've seen are `publish`, `pending`, `draft`, `auto-draft`, `trash`.
- GET `wc/v3/coupons?status=publish, draft, pending` would return coupons with status as `publish`, `pending` or `draft`.

This change would allow us to reliably show the status of the coupon when displaying the list of coupons.

## How to test the changes in this Pull Request:
I have added a separate class `class-wc-rest-coupons-controller-tests.php` to test the coupon controller class. For manual testing: 

1. Add a few coupons in wp-admin → Coupons → Add New.
2. Publish some of the coupons, update some coupons to draft or pending.
3. In the API, run `GET /wp-json/wc/v3/coupons` and confirm that all the coupons (regardless of the status) is fetched in the API response.
4. In the API, run `GET /wp-json/wc/v3/coupons?status=publish, draft, pending` and confirm that only coupons with those statuses are fetched in the API response.

## Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

## Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.

Adds the option to filter coupons by status when calling the `GET /coupons` endpoint. 

**This PR is in draft till [this PR](https://github.com/woocommerce/woocommerce/pull/31561) can be merged**

anitaa1990 committed:
Added the status field to the GET Coupon(s) API endpoint

anitaa1990 committed:
Updated existing unit test for reading coupons from the db

anitaa1990 committed:
Updated tests for version 2&3 when fetching coupons

anitaa1990 committed:
Merge branch 'trunk' into issue/31555-add-status-field-coupons-api

anitaa1990 committed:
Added option to filter coupons by status

anitaa1990 committed:
Added new unit test class for coupons v3 controller

anitaa1990 committed:
Added unit tests for filtering coupons by status

anitaa1990 committed:
Update plugins/woocommerce/includes/data-stores/class-wc-coupon-data-store-cpt.php

Co-authored-by: Josh Betz <j@joshbetz.com>

anitaa1990 committed:
Update plugins/woocommerce/includes/rest-api/Controllers/Version2/class-wc-rest-coupons-v2-controller.php

Co-authored-by: Josh Betz <j@joshbetz.com>

anitaa1990 committed:
Update plugins/woocommerce/includes/rest-api/Controllers/Version2/class-wc-rest-coupons-v2-controller.php

Co-authored-by: Josh Betz <j@joshbetz.com>

anitaa1990 committed:
Updated unit test for getters and setters

anitaa1990 committed:
Merge branch 'issue/31555-add-status-field-coupons-api' into issue/31562-filter-coupons-by-status

joshbetz commented on plugins/woocommerce/includes/rest-api/Controllers/Version3/class-wc-rest-coupons-controller.php:
```diff
@@ -24,4 +24,21 @@ class WC_REST_Coupons_Controller extends WC_REST_Coupons_V2_Controller {
 	protected $namespace = 'wc/v3';
+
+	protected function prepare_objects_query($request) {
+		$args = WC_REST_CRUD_Controller::prepare_objects_query( $request );
```
Since `WC_REST_Coupons_Controller` extends `WC_REST_Coupons_V2_Controller` which extends `WC_REST_CRUD_Controller`, we can use `$this` to reference the `prepare_objects_query` method.

```suggestion
		$args = $this->prepare_objects_query( $request );
```

joshbetz commented:
> GET /wc/v3/coupons would return all coupons, regardless of the status.

To maintain backwards compatibility, we may want to return only published coupons by default? We could keep everything else essentially the same, but add an `all` case for the status variable that returns all statuses.

anitaa1990 committed:
Update plugins/woocommerce/includes/rest-api/Controllers/Version3/class-wc-rest-coupons-controller.php

Co-authored-by: Josh Betz <j@joshbetz.com>

anitaa1990 committed:
Updated filter to support backward compatibility

anitaa1990 committed:
Merge branch 'issue/31562-filter-coupons-by-status' of https://github.com/woocommerce/woocommerce into issue/31562-filter-coupons-by-status

anitaa1990 committed:
Merge branch 'trunk' into issue/31562-filter-coupons-by-status

# Conflicts:
#	plugins/woocommerce/includes/data-stores/class-wc-coupon-data-store-cpt.php
#	plugins/woocommerce/tests/legacy/unit-tests/coupon/data.php

joshbetz commented on plugins/woocommerce/tests/legacy/unit-tests/coupon/data.php:
```diff
@@ -174,7 +174,7 @@ public function test_coupon_getters_and_setters() {
 			'discount_type'              => 'percent',
 			'amount'                     => 10.50,
-			'status'                     => 'publish',
+			'status'					 => 'publish',
```
@anitaa1990 I think the problem here is that there is a mix of tabs and spaces being used to align the `=>`. It should just be spaces.

```suggestion
		'status'                       => 'publish',
```

joshbetz committed:
Fix indentation

We use spaces to align the `=>` instead of a mix of tabs and spaces.

joshbetz committed:
Fix infinite loop

Oops. We can't call $this->prepare_objects_query() unconditionally inside
prepare_objects_query(). This was my fault 🙃

https://github.com/woocommerce/woocommerce/pull/31577/commits/b43af2ae16a580fb51db5ad446acbe2e6008559d

To explicitly reference the parent implementation, we use `parent::`

joshbetz committed:
Add doc comment for WC_REST_Coupons_Controller::prepare_objects_query

joshbetz reviewed (approved):
Looks good to me!
